pgmpy/base/ADMG.py

```diff
@@ -0,0 +1,410 @@
+import collections
+from typing import Hashable, Optional, Sequence
+import networkx as nx
+from networkx import MultiDiGraph
+
+from pgmpy.base.DAG import DAG as pgmpy_DAG
+
+
+class ADMG(MultiDiGraph):
+    """
+    Abstract class for an ADMG (Acyclic Directed Mixed Graph).
+    An ADMG is a directed graph that may contain both directed and undirected edges.
+    It is used to represent causal relationships in a system where some relationships are known to be directed,
+    while others are not specified as directed or undirected.
+    """
+
+    def __init__(self, directed_ebunch=None, bidirected_ebunch=None, latents=None):
+        super().__init__()
+        # Using edge attributes to distinguish bidirected edges
+        self.latents = set(latents) if latents else set()
+
+        if directed_ebunch:
+            self.add_directed_edges(directed_ebunch)
+        if bidirected_ebunch:
+            self.add_bidirected_edges(bidirected_ebunch)
+
+    def add_node(self, node):
+        super().add_node(node)
+
+    def add_nodes_from(self, nodes, **attr):
+        return super().add_nodes_from(nodes, **attr)
+
+    def _add_directed_edge(self, u, v):
+        if u is None or v is None:
+            raise ValueError("Can't add since one of nodes is None")
+
+        if u not in self.nodes:
+            self.add_node(u)
+        if v not in self.nodes:
+            self.add_node(v)
+
+        key = super().add_edge(u, v, type="directed")
+
+        if not nx.is_directed_acyclic_graph(self):
+            super().remove_edge(u, v, key=key)
+            raise ValueError("Adding this edge would create a cycle in the graph.")
+
+    def _add_bidirected_edge(self, u, v):
+        if u is None or v is None:
+            raise ValueError("Can't add since one of~ the nodes is None")
+        if u == v:
+            raise ValueError("Cannot add a bidirected edge from a node to itself.")
+
+        if u not in self.nodes:
```

Comment:
I don't think we need to include this as the `super().add_edge` would automatically add nodes if they do not exist.